### ISSUE TITLE
GuidGenerator can produce a Guid with embedded time of 10/15/1582

### ISSUE THREAD

**Issue by dlbromen** [fluentcassandra/fluentcassandra#117]
We've been using FluentCassandra v1.1.9's GuidGenerator in production now for a few months.  We generate 30+M Time UUIDs per day with it and it works great for the most part.  What I have noticed, is a handful of times per day, it will generate a Guid such that when I pass it into GuidGenerator.GetUtcDateTime, it returns the beginning of the Gregorian calendar, 10/15/1582.

A few other observations:
- The time of day component is not always the same.
- I cannot reproduce this with a multithreaded console app that is generating thousands of Guid per second.  I've let it run for a day straight and still nothing.

I just wanted to bring up this issue to:
1. See if anyone has a clue as to what conditions in the code would lead to this situation.
2. See if anyone else has seen this problem.
3. Start the resolution process, given there is a problem with the code somewhere.

**Comment by nberardi** [comment]
Have you logged those GUID's?

Also which method are you using?  Fast or NoDuplicates?

https://github.com/managedfusion/fluentcassandra/blob/master/src/GuidGenerator.cs#L181

**Comment by dlbromen** [comment]
Hi Nick,
I don't have them handy, but yes I did have a sampling.  I can either dig them up or get a new sample if that helps.

We're using the default (No Dups).

**Comment by nberardi** [comment]
Either works for me I just need to extract the ticks and see what is going on.  Is there anyway that you can record what they are before they go in and what they are after they come out.  Because I want to make sure that FluentCassandra is generating the faulty GUID or if it is a data corruption issue.

**Comment by dlbromen** [comment]
In the process of digging up some faulty ones.  Regarding, "Is there anyway that you can record what they are before they go in and what they are after they come out."  Not sure what you mean here or what you're after so I thought I'd ask so I can make sure I'm helping best I can.

**Comment by nberardi** [comment]
Lets start with the bad ones. As they come out of your database.

On Mar 4, 2013, at 2:44 PM, Derek Bromenshenkel notifications@github.com
wrote:

In the process of digging up some faulty ones. Regarding, "Is there anyway
that you can record what they are before they go in and what they are after
they come out." Not sure what you mean here or what you're after so I
thought I'd ask so I can make sure I'm helping best I can.

—
Reply to this email directly or view it on
GitHubhttps://github.com/managedfusion/fluentcassandra/issues/117#issuecomment-14400826
.

**Comment by dlbromen** [comment]
Guid be974d7c-0000-0000-0000-000000000000 - Date: 10/15/1582 12:05:19 AM - have seen this one more than once
Guid 8f9faa2a-0000-0000-0000-000000000000 - Date: 10/15/1582 12:04:00 AM

**Comment by dlbromen** [comment]
Another one that happened yesterday: Guid 4f432220-0000-0000-0000-000000000000 - Date: 10/15/1582 12:02:12 AM

**Comment by nberardi** [comment]
Were those generated by FluentCassandra?

## 

Nick Berardi
(484) 302-0125
Sent on the go from my phone.

On Mar 4, 2013, at 3:26 PM, Derek Bromenshenkel notifications@github.com
wrote:

Guid be974d7c-0000-0000-0000-000000000000 - Date: 10/15/1582 12:05:19 AM -
have seen this one more than once
Guid 8f9faa2a-0000-0000-0000-000000000000 - Date: 10/15/1582 12:04:00 AM

—
Reply to this email directly or view it on
GitHubhttps://github.com/managedfusion/fluentcassandra/issues/117#issuecomment-14403271
.

**Comment by nberardi** [comment]
To be honest these look like DateType values. Not TimeUuidType. If you were
to look at the bytes returned.

## 

Nick Berardi
(484) 302-0125
Sent on the go from my phone.

On Mar 4, 2013, at 3:33 PM, Derek Bromenshenkel notifications@github.com
wrote:

Another one that happened yesterday: Guid
4f432220-0000-0000-0000-000000000000 - Date: 10/15/1582 12:02:12 AM

—
Reply to this email directly or view it on
GitHubhttps://github.com/managedfusion/fluentcassandra/issues/117#issuecomment-14403643
.

**Comment by dlbromen** [comment]
Yes, they were generated by FluentCassandra.GuidGenerator.GenerateTimeBasedGuid().

I can find them by inspecting a string representation stored in a JSON object.  I don't actually look at the column name (time uuid type).

**Comment by nberardi** [comment]
I just don't understand how they could be generated. Because it is missing
the constant that says its a TimeUUID. Which it would be impossible not to
set.

Get the bytes out and check to see if it is an epoch. I have a suspicion
that something else is writing these.

## 

Nick Berardi
(484) 302-0125
Sent on the go from my phone.

On Mar 4, 2013, at 6:08 PM, Derek Bromenshenkel notifications@github.com
wrote:

Yes, they were generated by
FluentCassandra.GuidGenerator.GenerateTimeBasedGuid().

I can find them by inspecting a string representation stored in a JSON
object. I don't actually look at the column name (time uuid type).

—
Reply to this email directly or view it on
GitHubhttps://github.com/managedfusion/fluentcassandra/issues/117#issuecomment-14412767
.

**Comment by dlbromen** [comment]
Bytes? I'm just not clear on what that will gain us here.  Again I'm getting these from a JSON object.  At runtime we generate a TimeUUID and included it (string representation) in the JSON data and write the whole thing to a UTF8 column.

**Comment by nberardi** [comment]
What I am saying is that there is no way that this is a TimeUUID or that it was generated by FluentCassandra.  Lets break it down.

4f432220-0000-0000-0000-000000000000

The first 8 bytes or _4f432220-0000-0000_ is the timestamp from the gregorian calendar.  The next two bytes would be the clock sequence or _0000_ which is unlikely that it would ever be zero, given that you are using no-duplicates.  But there is the possibility.

The last 6 bytes or _000000000000_ are what I am saying is impossible and why this cannot be a TimeUUID, besides the impossibility of the rest.  Because the last 6 bytes are your mac address of the computer that generated them, and I doubt that it would ever be all zeros.  

Basically there was no way that GuidGenerater created this value. So something else in your architecture is inserting these values.  Or the data is becoming corrupted.  Because take a look at this method and tell me how those extra bytes could all be generated as zero so consistantly https://github.com/managedfusion/fluentcassandra/blob/master/src/GuidGenerator.cs#L239 especially since the last 6 bytes for the node should be the MAC address of the computer.  You would be seeing a whole bunch of these if that was the issue.

**Comment by nberardi** [comment]
If you find a path where you can generate a TimeUUID like that please let me know, because it shouldn't be happening, but I am very confident in this code and the simple process that it goes through.

**Comment by dlbromen** [comment]
I completely agree this is odd.  But, we literally are talking 1-in-a-million here - it works flawlessly millions of times per day.

The code path for this in the app is simple.  I can tell this is happening at generation-time _before_ persisting it anywhere.  I am certain FluentCassandra generated them.

I think it is a mistake to dismiss this simply because it appears impossible.  I'm thinking along the lines of a race condition, hardware glitch, .NET framework bug (Stopwatch, PhysicalAddress, etc).

You're going down the road I was hoping for - analyzing how the thing is constructed because you understand it best. I think we can go more with some "what-if" scenarios.  What if the clock sequence or mac address were 0s (for whatever impossible reason)?  Do the first 8 bytes even make sense? They are suspicious too because the end of it has a lot of consecutive zeros.

**Comment by nberardi** [comment]
Well I don't have an answer. I guess the only thing you could do is create
a program that generates millions of instances in hopes of reproducing.

## 

Nick Berardi
(484) 302-0125
Sent on the go from my phone.

On Mar 4, 2013, at 8:34 PM, Derek Bromenshenkel notifications@github.com
wrote:

I completely agree this is odd. But, we literally are talking
1-in-a-million here - it works flawlessly millions of times per day.

The code path for this in the app is simple. I can tell this is happening
at generation-time _before_ persisting it anywhere. I am certain
FluentCassandra generated them.

I think it is a mistake to dismiss this simply because it appears
impossible. I'm thinking along the lines of a race condition, hardware
glitch, .NET framework bug (Stopwatch, PhysicalAddress, etc).

You're going down the road I was hoping for - analyzing how the thing is
constructed because you understand it best. I think we can go more with
some "what-if" scenarios. What if the clock sequence or mac address were 0s
(for whatever impossible reason)? Do the first 8 bytes even make sense?
They are suspicious too because the end of it has a lot of consecutive
zeros.

—
Reply to this email directly or view it on
GitHubhttps://github.com/managedfusion/fluentcassandra/issues/117#issuecomment-14418540
.

**Comment by dlbromen** [comment]
I have tried that to some extent as mentioned in the original post.  The challenge I pose is: Let's say we could reproduce it once in 24 hours while generating these.  Then what?  What do we capture?  

The reason I say this, is to a degree I am reproducing it every day in production.  Depending on what the next step is, I may be able to do that in production.

**Comment by nberardi** [comment]
Well lets first track down the server whose Node is set to 000000000000, because that one is probably the culprit since the Node is only set once.

**Comment by dlbromen** [comment]
According to our server logs, all of them regularly experience this problem.  It appears it is just one bad GUID at a time, though. What I mean is a given server can have multiple faults over time, but never a whole bunch of them back-to-back.

**Comment by nberardi** [comment]
Which is odd because the node is set once and never changed.

My I ask a question if you are just serializing a TimeUUID down to a UTF8.  Why not just store a DateTimeOffset.Now?  Or store the ticks in a IntType?

**Comment by dlbromen** [comment]
We are also using it the "useful" way where it is a uuid column name type
and we're storing ordered, unique columns. I just didn't want to distract
the conversation with this and other complicated parts of the system.
I just want to make sure it is understood that this is happening at
generation time.

**Comment by nberardi** [comment]
My hands are sort of tied until I can reproduce this behavior that you describe.  

Because here is what would have to go wrong for this to happen.

line 253 has to get converting the types wrong from the ticks
lines 256, 259, 265, 266, 269, and 270 need to be skipped

in this method https://github.com/managedfusion/fluentcassandra/blob/master/src/GuidGenerator.cs#L237

I can't easily reproduce that behavior.  I have even explored the possibility that the byte arrays used for node and clock sequence were some how reset to zero for a split second.  that would account for the last set of zeros.  However it still would require lines 265, 266, 269, and 270 to be skipped.  These things cannot happen.

**Comment by nberardi** [comment]
So where did we land on this one?

**Comment by dlbromen** [comment]
It is still happening tens or hundreds of times per day in our system.  It got put on the back burner after I failed to reproduce it with a test app and an extremely simplified use case.

We haven't dropped it though, and at some point we'll get around to looking at it again.

**Comment by nberardi** [comment]
Ok I would be interested and seeing where this is coming from, because nothing I can see right now could generate what you are describing, but that doesn't mean that something isn't happening in the code.

**Comment by dlbromen** [comment]
I finally got back around to investigating this.  At this point, I am 99% sure I have discovered the problem, and ruled out FluentCassandra as the cause.  For that reason, I am closing this issue.

For the good of the community, I will explain what I found in more detail once I have confirmation in production that I have resolved the issue.  Long story short is: Assignment of a Guid variable is not atomic or thread-safe, and in just the right conditions you can experience a "torn read" where _some_ of the in-memory components of the Guid struct are still the 'old' value (eg, 0s in an empty guid), while others are the new value from the in-progress assignment (eg, result of FC generation).

Thanks to Nick for both hearing me out and challenging me to dig deeper along the way.
